src/ucp/core/ucp_am.c

```diff
@@ -668,38 +690,54 @@ static ucs_status_t ucp_am_zcopy_single(uct_pending_req_t *self)
 static ucs_status_t ucp_am_zcopy_single_reply(uct_pending_req_t *self)
 {
     ucp_request_t *req = ucs_container_of(self, ucp_request_t, send.uct);
-    ucp_am_reply_hdr_t reply_hdr;
+    ucp_am_hdr_t hdr;
+    ucp_am_reply_ftr_t *ftr;
 
-    reply_hdr.ep_id = ucp_send_request_get_ep_remote_id(req);
-    ucp_am_fill_header(&reply_hdr.super, req);
+    ucp_am_fill_header(&hdr, req);
+    ucs_assert(req->send.msg_proto.am.reg_desc != NULL);
 
-    return ucp_do_am_zcopy_single(self, UCP_AM_ID_SINGLE_REPLY, &reply_hdr,
-                                  sizeof(reply_hdr),
-                                  req->send.msg_proto.am.reg_desc,
-                                  req->send.msg_proto.am.header_length,
+    ftr        = UCS_PTR_BYTE_OFFSET(req->send.msg_proto.am.reg_desc + 1,
+                                     req->send.msg_proto.am.header_length);
+    ftr->ep_id = ucp_send_request_get_ep_remote_id(req);
+
+    return ucp_do_am_zcopy_single(self, UCP_AM_ID_SINGLE_REPLY, &hdr,
+                                  sizeof(hdr), req->send.msg_proto.am.reg_desc,
+                                  req->send.msg_proto.am.header_length +
+                                          sizeof(*ftr),
```

Comment:
should we fix alignment here?

Reply:
clang likes this way, because it is a part of arithmetic operation, but not a new parameter